### PR TITLE
feat(vrl): Adds `chunks` function

### PR DESCRIPTION
**Pull request by briankung**
Issue: https://github.com/vectordotdev/vector/issues/13329

This adds a `chunks` method (chunks rather than chunk following the slice naming convention for [slice's chunks method](https://doc.rust-lang.org/std/primitive.slice.html#method.chunks)). This chunks data based on a given `chunk_size` in bytes. The result is an array of bytes of `chunk_size` in length.

______
👇 The below was true for the initial PR, which attempted to respect unicode code point boundaries. After [this comment](https://github.com/vectordotdev/vector/pull/13794#pullrequestreview-1065676560) that is no longer the aim of this PR.
______

~It attempts to respect unicode code point boundaries, but makes no guarantees about grapheme cluster boundaries: https://unicode.org/reports/tr29/#Grapheme_Cluster_Boundaries~

~A quick overview is that valid UTF-8 code points are up to 4 bytes in length, the length of which is denoted by the first few bits in the first byte. See https://en.wikipedia.org/wiki/UTF-8#Encoding).~

~However, *user-perceived characters* can consist of several UTF-8 code points (see above link on Grapheme Cluster Boundaries). There is no simple algorithm for splitting text based on grapheme cluster boundaries because grapheme clusters are of arbitrary lengths.~

~Therefore, the algorithm used here only attempts to split text based on code point boundaries, which are at most 4 bytes in length, and not on grapheme cluster boundaries.~

Additional subtasks that I would like to complete:
- [x] Basic functionality - splitting bytes by `chunk_size`
- [x] Respects unicode code point boundaries 
- [x] Fix `fn examples` tests
- [x] ~Consider refactoring the main loop~
- [x] Add a `utf` configuration flag to treat it as utf or not (default: true)
- [x] Add documentation
    - [x] Mention the unicode code points limitation
    - [x] Mention the chunk_size limitation
- [x] ~~Add tests for data size (i.e. check that a 1MB limit actually does limit it to 1MB chunks)~~ I don't want to add 1MB+ to the repo, that would be silly

### CODE REVIEW

File: lib/value/src/value/convert.rs [hunk 1]
```diff
@@ -264,6 +264,12 @@ impl<const N: usize> From<&[u8; N]> for Value {
     }
 }
 
+impl From<&[u8]> for Value {
+    fn from(data: &[u8]) -> Self {
+        Self::from(Bytes::copy_from_slice(data))
+    }
+}
+
```
Comment (briankung) [inline lib/value/src/value/convert.rs:272]: I added this to fit the type signature of the code, but in general I think I have a lot to learn about the Vector types. Would appreciate feedback on whether this is necessary or if there's a better way to use the existing trait implementations, given the code in the `chunks` method

File: lib/vrl/stdlib/Cargo.toml [hunk 2]
```diff
@@ -75,6 +75,7 @@ default = [
     "assert_eq",
     "boolean",
     "ceil",
+    "chunks",
```
Comment (briankung) [inline lib/vrl/stdlib/Cargo.toml:78]: I'm assuming this should be a default feature, but please let me know if this is not the case.

File: lib/vrl/stdlib/src/chunks.rs [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            backtrack = 0;
+            start = end;
+        } else {
+            backtrack += 1;
+
+            if backtrack >= MAX_UTF8_CODE_POINT_LENGTH_IN_BYTES {
+                return Err("Bytes are not valid UTF-8".into());
+            }
+        }
+    }
```
Comment (briankung) [inline lib/vrl/stdlib/src/chunks.rs:42]: This is a very imperative way to chunk the data, but it made the most sense at the time. I would appreciate feedback on other - especially more functional - ways to approach it.

File: lib/vrl/stdlib/src/chunks.rs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            tdef: TypeDef::array(Collection::from_unknown(Kind::bytes())),
+        }
+
+        invalid_utf8 {
+            args: func_args![value: b"\xa0\xa1\xa2\xa3\xa4\xa5\xa6\xa7\xa8\xa9\xaa\xab\xac\xad\xae\xaf\xb0",
+                             chunk_size: 8
+            ],
+            want: Err("Bytes are not valid UTF-8"),
+            tdef: TypeDef::array(Collection::from_unknown(Kind::bytes())),
+        }
```
Comment (briankung) [inline lib/vrl/stdlib/src/chunks.rs:151]: This test macro is cool, but I think I need to learn more about the `tdef` parameter and valid values for it.

File: lib/vrl/stdlib/src/chunks.rs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        let value = self.value.resolve(ctx)?;
+        let chunk_size = self.chunk_size.resolve(ctx)?;
+
+        chunks(value, chunk_size)
+    }
+
+    fn type_def(&self, _state: (&state::LocalEnv, &state::ExternalEnv)) -> TypeDef {
+        TypeDef::array(Collection::from_unknown(Kind::bytes())).infallible()
+    }
+}
```
Comment (briankung) [inline lib/vrl/stdlib/src/chunks.rs:112]: I'd love to know what `impl Expression` and `impl Function` do in the Vector codebase

**Comment by briankung** [comment]
> # Soak Test Results
> 
> Baseline: [e5c435f](https://github.com/vectordotdev/vector/commit/e5c435f8c7049fea93438408059f02167c61b747) Comparison: [00ef727](https://github.com/vectordotdev/vector/commit/00ef7272a769c2e947f62919e7071dd945de675a) Total Vector CPUs: 4
> Explanation
> 
> Changes in throughput with confidence ≥ 90.00% and absolute Δ mean >= ±8.87%:
> experiment 	Δ mean 	Δ mean % 	confidence
> socket_to_socket_blackhole 	-9.08MiB 	-40.2 	100.00%
> Fine details of change detection per experiment.

Looks like I just had to rebase.

_____

- ❓ How can I run the `fn examples` code in the `Function` trait impl locally? I tried to run `make test-vrl` but it looked like it hung while trying to update a git submodule.

**Comment by tobz** [comment]
> ❓ How can I run the fn examples code in the Function trait impl locally? I tried to run make test-vrl but it looked like it hung while trying to update a git submodule.

That would indeed be the way to do it, but it's surprising that building Vector itself didn't already get you all of the dependencies needed to build the `test-vrl` binary.  We can try and help if you can show us the output where it's hanging.

**Comment by briankung** [comment]
@tobz seems like it's working now, whatever it was! Maybe I just had a connectivity issue. Being able to run the tests helps a lot ^_^

File: lib/vrl/stdlib/src/chunks.rs [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        chunks(value, chunk_size, utf8)
+    }
+
+    fn type_def(&self, _state: (&state::LocalEnv, &state::ExternalEnv)) -> TypeDef {
+        TypeDef::array(Collection::from_unknown(Kind::bytes())).infallible()
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (jszwedko) [inline lib/vrl/stdlib/src/chunks.rs:170]: In addition to these tests, we usually add benchmarks for each new VRL function, as well, in `benches/benches.rs`.

File: lib/vrl/stdlib/src/chunks.rs (hunk 5 above)
Comment (jszwedko) [reply lib/vrl/stdlib/src/chunks.rs:167]: These are traits required by the compiler for each function. `Function` describes the function, and then `Expression` implements the evaluation logic.

**Review by jszwedko — COMMENTED** [review]
Thanks for this contribution @briankung !

It prompted some discussion on the team about the inconsistencies with how VRL deals with bytes and UTF-8 strings (some of which are described in https://github.com/vectordotdev/vector/issues/12369). We think the best plan is to follow up with an RFC document to lay out a direction in the future, but, in absence of that, we think the simplest thing to do for this function would actually be to just have `chunks` operate on the number of bytes without any specific UTF-8 handling (that is, remove the `utf8` flag). Then we can follow up later once we have defined some guidelines for how functions in VRL deal with UTF-8 strings and bytes to make sure all VRL functions handle them in a consistent fashion.

Other than that, what you have here generally looks good to me. I left one comment below about adding benchmarks.

File: lib/vrl/stdlib/src/chunks.rs [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?;
+        let chunk_size = self.chunk_size.resolve(ctx)?;
+        let utf8 = self.utf8.resolve(ctx)?;
+
+        chunks(value, chunk_size, utf8)
+    }
+
+    fn type_def(&self, _state: (&state::LocalEnv, &state::ExternalEnv)) -> TypeDef {
+        TypeDef::array(Collection::from_unknown(Kind::bytes())).infallible()
```
Comment (fuchsnj) [inline lib/vrl/stdlib/src/chunks.rs:165]: This function is fallible, for example if the chunk size is set too low.
```return Err(r#""chunk_size" must be at least 1 byte"#.into());```

Since it's likely that a literal number will be passed into the function, we can be a bit smarter though. You can call `Expression::as_value` on the `chunk_size` argument, and this will resolve to a value at compile time if it can be calculated (if it's a literal). If so, you can do the length check at compile time and if it's >=1, return `infallible` here. Otherwise, it needs to be `fallible` since it could potentially fail.

You can assume the argument types will be correct when calculating fallibility here, since that happens automatically by the compiler.

File: lib/vrl/stdlib/src/chunks.rs (hunk 4 above)
Comment (fuchsnj) [reply lib/vrl/stdlib/src/chunks.rs:219]: `tdef` is the type definition that this function resolves to. In VRL, all expressions resolve to a value, and the type definition describes the possible values that it can resolve to, as well as whether it is fallible or not..

In this case,  `TypeDef::array(Collection::from_unknown(Kind::bytes()))` is just saying the function returns an array of _any_ size, where the elements must be `bytes`.

File: lib/vrl/stdlib/src/chunks.rs (hunk 7 above)
Comment (briankung) [reply lib/vrl/stdlib/src/chunks.rs:165]: What happens when you use `Expression:as_value` on `chunk_size` but the `chunk_size` parameter is not a literal?

Is marking it as fallible or infallible just for documentation purposes? i.e. what are the ramifications of incorrectly annotating its fallibility?

Comment (fuchsnj) [reply lib/vrl/stdlib/src/chunks.rs:165]: >What happens when you use Expression:as_value on chunk_size but the chunk_size parameter is not a literal?

`as_value()` returns an `Option<Value>`. It will return `None` if the value is not known at compile time.

> Is marking it as fallible or infallible just for documentation purposes? i.e. what are the ramifications of incorrectly annotating its fallibility?

This affects the type definition of the function, which is used at compile time. Fallible functions must have some kind of error handling.

I'd suggest running `vector vrl` if you haven't already to get access to the VRL REPL. You can play around with the function there and see how changing the type definition affects things.

Comment (briankung) [reply lib/vrl/stdlib/src/chunks.rs:165]: > I'd suggest running `vector vrl` if you haven't already to get access to the VRL REPL. You can play around with the function there and see how changing the type definition affects things.

Oh that sounds cool, I'll try that, thanks! 👍

Comment (briankung) [reply lib/vrl/stdlib/src/chunks.rs:165]: Oh okay, after playing with the REPL everything makes a whole lot more sense 😅 thanks again!

I'm trying to make it infallible with code like this:

```diff
diff --git a/lib/vrl/stdlib/src/chunks.rs b/lib/vrl/stdlib/src/chunks.rs
index 88afb3826..5872413bf 100644
--- a/lib/vrl/stdlib/src/chunks.rs
+++ b/lib/vrl/stdlib/src/chunks.rs
@@ -5,10 +5,6 @@ fn chunks(value: Value, chunk_size: Value) -> Resolved {
     let chunk_size = chunk_size.try_integer()? as usize;
     let bytes = value.try_bytes()?;
 
-    if chunk_size < 1 {
-        return Err(r#""chunk_size" must be at least 1 byte"#.into());
-    }
-
     Ok(bytes.chunks(chunk_size).collect::<Vec<_>>().into())
 }
 
@@ -66,6 +62,17 @@ impl Function for Chunks {
         let value = arguments.required("value");
         let chunk_size = arguments.required("chunk_size");
 
+        if let Some(chunk_size) = chunk_size.as_value() {
+            if chunk_size < 1 {
+                return Err(vrl::function::Error::InvalidArgument {
+                    keyword: "chunk_size",
+                    value: chunk_size,
+                    error: "Invalid chunk_size",
+                }
+                .into());
+            };
+        };
+
         Ok(Box::new(ChunksFn { value, chunk_size }))
     }
 }
```

But I'm getting this compiler error:

```
   Compiling vrl-stdlib v0.1.0 (/Users/brian/Code/github/vectordotdev/vector/lib/vrl/stdlib)
error[E0308]: mismatched types
  --> lib/vrl/stdlib/src/chunks.rs:66:29
   |
66 |             if chunk_size < 1 {
   |                             ^ expected enum `value::Value`, found integer
   |
help: try wrapping the expression in `value::Value::Integer`
   |
66 |             if chunk_size < value::Value::Integer(1) {
   |                             ++++++++++++++++++++++ +

For more information about this error, try `rustc --explain E0308`.
error: could not compile `vrl-stdlib` due to previous error
```
I'd like to be able to mark this function as infallible, but I'm not sure where to go from here (and I've got an errand to run). If you could provide some guidance on how to compare the Value to the integer, that would help! Otherwise I may just mark the function as fallible.

Comment (fuchsnj) [reply lib/vrl/stdlib/src/chunks.rs:165]: > if let Some(chunk_size) = chunk_size.as_value() {

This gives you a `Value`. A `Value` is an enum of the different runtime types. You can check if the actual argument is an integer / get that integer by calling `as_integer` which returns `Option<i64>`, and then compare that value to 1.

Comment (fuchsnj) [reply lib/vrl/stdlib/src/chunks.rs:165]: You will still want to keep the runtime check for the `chunk_size` that you deleted in the diff. This is useful if the argument passed in is a variable, or something unknown at compile time (which could fail)

File: lib/vrl/stdlib/src/chunks.rs [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+fn chunks(value: Value, chunk_size: Value) -> Resolved {
+    let chunk_size = chunk_size.try_integer()?;
+    let bytes = value.try_bytes()?;
+
+    if chunk_size < 1 {
+        return Err(r#""chunk_size" must be at least 1 byte"#.into());
+    }
+
+    Ok(bytes.chunks(chunk_size as usize).collect::<Vec<_>>().into())
```
Comment (briankung) [inline lib/vrl/stdlib/src/chunks.rs:12]: In previous commits, I was casting `chunk_size` to `usize` in the `let` expression, but it led to the discovery that negative numbers were being cast to `usize::MAX` (see this issue: https://github.com/vectordotdev/vector/issues/13891). Casting it to `usize` here means allows the chunk size check to occur on [L9](https://github.com/vectordotdev/vector/pull/13794/files#diff-2166716f04399fa4177d788683336e029a5e889207341eca092761ff5a585e1dR9) before being cast to `usize`.

However, it does raise a question: will vector always be compiled on a 64-bit target architecture? e.g. 32-bit architectures would have a smaller implicit maximum `chunk_size`.

File: lib/vrl/stdlib/src/chunks.rs (hunk 7 above)
Comment (briankung) [reply lib/vrl/stdlib/src/chunks.rs]: @fuchsnj I spent a bit of time trying to make the function infallible - I haven't gotten it quite yet, but my current direction is to try to enforce that the expression is a literal and not an expression similar to [redact.rs#L101-L139](https://github.com/vectordotdev/vector/blob/master/lib/vrl/stdlib/src/redact.rs#L101-L139). Does that sound like the right way to go?

I'm also checking in `fn compile()` to see if the given value is `>=1`:

```rust
if let Some(cs_value) = chunk_size.as_value() {
    if let Some(cs_int) = cs_value.as_integer() {
        if cs_int < 1 {
            return Err(vrl::function::Error::InvalidArgument {
                keyword: "chunk_size",
                value: cs_value,
                error: r#""chunk_size" must be at least 1 byte"#,
            }
            .into());
        }
    }
};
```

Comment (briankung) [reply lib/vrl/stdlib/src/chunks.rs]: I'm also a bit torn on whether this should be fallible or infallible. More thoughts on that when I can, but input would be appreciated, as well.

File: lib/vrl/stdlib/src/chunks.rs (hunk 8 above)
Comment (fuchsnj) [reply lib/vrl/stdlib/src/chunks.rs:12]: Thanks for checking this!

Vector can be compiled for 32-bit targets. You can use `usize::try_from` to see if a value can be correctly converted. It should also be added to the type definition, so valid literals won't make the function fallible. You can also do a check with `usize::BITS` and if it's >=64, the check isn't necessary even if the `chunk_size` isn't a literal.

**Comment by briankung** [comment]
I think this is ready for final review. I have two minor questions regarding testing:

1. Should I write tests for cases where `chunks` is fallible? e.g. when it's been supplied with an expression for `chunk_size` or when `chunk_size` doesn't fit into a `usize`
2. If so, is there a way to write a test to be run on a particular architecture? On 32-bit architectures, it's possible for the `chunk_size` parameter to be constrained by `usize::MAX`.

There are currently no test cases covering these edge cases.

File: lib/vrl/stdlib/src/chunks.rs [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+    let chunk_size = chunk_size.try_integer()?;
+
+    if chunk_size < 1 {
+        return Err(r#""chunk_size" must be at least 1 byte"#.into());
+    }
+
+    if let Ok(chunk_size) = usize::try_from(chunk_size) {
+        Ok(bytes.chunks(chunk_size).collect::<Vec<_>>().into())
+    } else {
+        Err(r#""chunk_size" must be a valid usize for this target architecture"#.into())
```
Comment (fuchsnj) [inline lib/vrl/stdlib/src/chunks.rs:15]: The term `usize` may be confusing to VRL users who aren't familiar with Rust. Maybe re-word this to just say the value is out of the valid range (and maybe calculate / display the actual valid range)

File: lib/vrl/stdlib/src/chunks.rs [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+impl Expression for ChunksFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?;
+        let chunk_size = self.chunk_size.resolve(ctx)?;
+
+        chunks(value, chunk_size)
+    }
+
+    fn type_def(&self, _state: (&state::LocalEnv, &state::ExternalEnv)) -> TypeDef {
+        let chunk_size_is_expression = self.chunk_size.as_value().is_none();
```
Comment (fuchsnj) [inline lib/vrl/stdlib/src/chunks.rs:92]: This is close, but it currently assuming _any_ literal value is infallible. You still need to check the actual literal value (eg: A literal `0` is fallible).

File: lib/vrl/stdlib/src/chunks.rs [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+        // chunk_size is converted to a usize later, so if a user-supplied Value::Integer (i64) is
+        // larger than the platform's usize::MAX, it could fail to convert.
+        let chunk_size_limited_by_architecture = usize::BITS < 64;
```
Comment (fuchsnj) [inline lib/vrl/stdlib/src/chunks.rs:95]: Currently if a _valid_ literal is passed in on a 32-bit system, the function will be marked infallible. The `usize::BITS` value should probably only be used when checking if a literal is in the valid range. If you don't have a literal, the function should already be marked as `infallible` (since we don't know what the input could be)

File: lib/vrl/stdlib/src/chunks.rs [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+            want: Ok(value!(["abcd", "efgh"])),
+            tdef: TypeDef::array(Collection::from_unknown(Kind::bytes())),
+        }
+
+        minimum_chunk_size {
+            args: func_args![value: "",
+                             chunk_size: 0,
+            ],
+            want: Err(r#""chunk_size" must be at least 1 byte"#),
+            tdef: TypeDef::array(Collection::from_unknown(Kind::bytes())),
```
Comment (fuchsnj) [inline lib/vrl/stdlib/src/chunks.rs:122]: This is an invalid type definition. (The function had an error, but the type definition is infallible)

File: lib/vrl/stdlib/src/chunks.rs (hunk 11 above)
Comment (briankung) [reply lib/vrl/stdlib/src/chunks.rs:95]: > If you don't have a literal, the function should already be marked as `infallible` (since we don't know what the input could be)

I may be misunderstanding what `infallible` means, but do you mean `fallible` here?

Comment (fuchsnj) [reply lib/vrl/stdlib/src/chunks.rs:95]: > I may be misunderstanding what infallible means, but do you mean fallible here?

Yep, sorry! I fixed it.

Comment (briankung) [reply lib/vrl/stdlib/src/chunks.rs:95]: So if the user supplies a literal `chunk_size` that we know is not within the valid range, can we say definitively at that point that `chunks` will fail? As a user, I think I would prefer it to be a compile time error rather than handling the fallibility and then also discovering the error at runtime, especially when it's something like 0 or \*checks notes\* 4294967296 (~4GiB).

As a side note, I think using `usize::BITS` may be a bit of a red herring since the easiest way to see if the chunk size is within a valid range is to use `usize::try_from(chunk_size)` and `chunk_size > 0`. `usize::BITS` tells us about the platform's `usize`, but we can just check if it's a valid `usize` directly.

My proposal:
1. Make it fallible if it's an expression
2. Make it a compile time error if it's an out-of-range literal
3. Else make it infallible

I think the compile time checks would be done in `fn compile`, right? However from our call, I got the impression that that compile time checks on the literal values aren't generally done that way? Let me know what you think.

Comment (fuchsnj) [reply lib/vrl/stdlib/src/chunks.rs:95]: > So if the user supplies a literal chunk_size that we know is not within the valid range, can we say definitively at that point that chunks will fail? As a user, I think I would prefer it to be a compile time error rather than handling the fallibility and then also discovering the error at runtime

That's reasonable. I'm fine with failing with a compile time error if it's guaranteed to fail.

> As a side note, I think using usize::BITS may be a bit of a red herring since the easiest way to see if the chunk size is within a valid range is to use usize::try_from(chunk_size) and chunk_size > 0. usize::BITS tells us about the platform's usize, but we can just check if it's a valid usize directly.

Right. Using `usize::try_from` should be sufficient.

Overall your proposal is reasonable, although I would re-word 1 to "Make it fallible if it's not a literal" (since _all_ arguments are expressions)

> I think the compile time checks would be done in fn compile, right? However from our call, I got the impression that that compile time checks on the literal values aren't generally done that way? Let me know what you think.

Correct. We don't do this today in many places (for example `1/0` compiles, but is fallible), but I'm not opposed to adding it. I'm considering expanding the "fallibility" concept in the future to be an enum of something like `infallible` / `fallible` / `always_fails`, and then the compiler can automatically error if it's guaranteed to fail, but for now adding it in the `compile` function is fine.

Comment (briankung) [reply lib/vrl/stdlib/src/chunks.rs]: > I'm considering expanding the "fallibility" concept in the future to be an enum of something like `infallible` / `fallible` / `always_fails`, and then the compiler can automatically error if it's guaranteed to fail

I think this would be helpful in this case, because then I could check whether `chunk_size` is in range in one place instead of 3.

File: lib/vrl/stdlib/src/chunks.rs (hunk 12 above)
Comment (briankung) [reply lib/vrl/stdlib/src/chunks.rs:139]: This seems to pass either way in the latest commit. Since it's now a compile time error in ce0e9b6ee6d0ed5000108b1cb31137a56568e8ed, should it be marked as fallible?

I considered taking this test out as well since I covered it in `chunks_chunk_size_at_least_1.vrl`, since it's a compile time error.

Comment (fuchsnj) [reply lib/vrl/stdlib/src/chunks.rs:139]: I'd just remove it from here since it's covered in the VRL test already. I don't think these tests handle compile time errors well yet.

File: lib/vrl/stdlib/src/chunks.rs [hunk 13]
```diff
@@ -0,0 +1,151 @@
+use ::value::Value;
+use vrl::prelude::*;
+
+fn chunks(value: Value, chunk_size: Value) -> Resolved {
+    let bytes = value.try_bytes()?;
+    let chunk_size = chunk_size.try_integer()?;
+
+    if chunk_size < 1 {
+        return Err(r#""chunk_size" must be at least 1 byte"#.into());
```
Comment (fuchsnj) [inline lib/vrl/stdlib/src/chunks.rs:9]: This message uses a `byte` unit, where the one below is just a number. It'd be nice to keep them consistent.

File: lib/vrl/stdlib/src/chunks.rs [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ChunksFn {
+    value: Box<dyn Expression>,
+    chunk_size: Box<dyn Expression>,
+}
+
+impl Expression for ChunksFn {
```
Comment (fuchsnj) [inline lib/vrl/stdlib/src/chunks.rs:104]: I merged some changes into `master` recently that modifies how functions are implemented. Please switch this from `impl Expression ...` to `impl FunctionExpression`.

File: lib/vrl/stdlib/src/chunks.rs [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+                        keyword: "chunk_size",
+                        value: literal,
+                        error: r#""chunk_size" is too large"#,
+                    }
+                    .into());
+                }
+            }
+        }
+
+        Ok(Box::new(ChunksFn { value, chunk_size }))
```
Comment (fuchsnj) [inline lib/vrl/stdlib/src/chunks.rs:94]: After implementing `FunctionExpression`, this needs to change a bit (see comment below)
```suggestion
        Ok(ChunksFn { value, chunk_size }.as_expr())
```

File: lib/vrl/stdlib/src/chunks.rs (hunk 14 above)
Comment (fuchsnj) [inline lib/vrl/stdlib/src/chunks.rs:104]:
```suggestion
impl FunctionExpression for ChunksFn {
```

File: lib/vrl/stdlib/src/chunks.rs [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+
+impl Expression for ChunksFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?;
+        let chunk_size = self.chunk_size.resolve(ctx)?;
+
+        chunks(value, chunk_size)
+    }
+
+    fn type_def(&self, _state: (&state::LocalEnv, &state::ExternalEnv)) -> TypeDef {
```
Comment (fuchsnj) [inline lib/vrl/stdlib/src/chunks.rs:112]:
```suggestion
    fn type_def(&self, _state: &TypeState) -> TypeDef {
```

**Review by fuchsnj — COMMENTED** [review]
A few small things / merge conflicts. Just need to resolve those and fix some of the CI issues and this should be good to go.

**Comment by briankung** [comment]
Rebased and updated!

**Comment by briankung** [comment]
I can't tell why the cue docs are failing in CI. Locally, I get:

```
$ CI=true make check-docs
./scripts/check-docs.sh
Validating cue files formatting...
Validating cue files correctness...
Success! The contents of the sources in the "./website/cue" directory are valid
```

**Review by fuchsnj — APPROVED** [review]

**Comment by fuchsnj** [comment]
@briankung Thanks for the contribution!

**Comment by briankung** [comment]
Thanks for the multiple rounds of reviews! Excited to see this in a new release.
